src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,795 @@
-                            timely::PartialOrder::less_equal(&self.read_progress, &progress),
-                            "{:?} PARTIAL ORDER: {:?} {:?}",
-                            self.name,
-                            self.read_progress,
-                            progress
-                        );
-                        self.read_progress = progress;
-                    }
-                    ListenEvent::Updates(updates) => {
-                        for ((_, value), timestamp, diff) in updates {
-                            let partition = value.expect("Unable to decode partition id");
-                            self.persisted_timestamp_bindings
-                                .entry(partition)
-                                .or_insert_with(VecDeque::new)
-                                .push_back((timestamp, diff));
-                        }
-                    }
+                    Ok(Ok(Ok(()))) => break,
+                    Ok(Ok(Err(actual_upper))) => return Err(actual_upper),
+                    Ok(Err(invalid_use)) => panic!("{invalid_use:?}"),
+                    Err(_external_err) => continue,
                 }
             }
+            // At this point we have successfully produced data in the reclock shard so we need to
+            // sync to update our local view as well
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
+
+    /// Produces a new timestamp suitable for minting bindings or the amount of time that the
+    /// caller needs to wait for one to become available. On success, the returned timestamp is
+    /// guaranteed to be beyond the current `upper` frontier and a multiple of `update_interval_ms`
+    fn next_mint_timestamp(&self) -> Result<Timestamp, Duration> {
+        let now = (self.now)();
+        let mut new_ts = now - now % self.update_interval_ms;
+        if (now % self.update_interval_ms) != 0 {
+            new_ts += self.update_interval_ms;
+        }
+        let upper_ts = self.upper.as_option().expect("no more timestamps to mint");
+        if upper_ts <= &new_ts {
+            Ok(new_ts)
+        } else {
+            Err(Duration::from_millis(upper_ts - now))
+        }
+    }
+
+    /// Returns an iterator of timestamp bindings for a given partition
+    fn partition_uppers(&self, pid: &PartitionId) -> PartitionUppers {
+        let bindings = match self.remap_trace.get(pid) {
+            Some(bindings) => (*bindings).iter(),
+            None => (&[]).iter(),
+        };
+        PartitionUppers {
+            offset: MzOffset::default(),
+            bindings,
+        }
+    }
 
-            debug_assert!(self.validate_persisted_bindings());
+    /// Returns the since frontier for a given partition
+    fn partition_since(&self, pid: &PartitionId) -> MzOffset {
+        if self.since.elements() == [Timestamp::minimum()] {
+            // If we never compacted in the DestTime domain then the SourceTime domain isn't
```

Comment:
The operator knows if a compaction has happened or not by checking if the since frontier is `Antichain::from_elem(Timestamp::minimum())` or not. When the operator is started with an `as_of` frontier then the since frontier is initialized as such because during the initial snapshot [persist will *compact itself*. ](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/read.rs#L138)

In other words there is no material difference between:
1. starting the operator `as_of 0` and then compacting to timestamp `t`
2. starting the operator `as_of t`

The final state is indistinguishable and both should be treated as compaction having happened.

I will add a test that uses `as_of` initialization nevertheless to exercise that code path too.

Reply:
Having messages reclocked *at* the `as_of` is not possible. In order to allow it I would have to remove the safety feature if refusing to reclock messages that are in the past (and might be reclocked incorrectly). The compaction/as_of test that is in place already tests what happens at the boundary